nvflare/app_common/abstract/statistics_spec.py

```diff
@@ -0,0 +1,224 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+from typing import Dict, List
+
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_context import FLContext
+from nvflare.app_common.abstract.statistics_def import Feature, Histogram
+
+"""
+    Statistics defines methods that user need to implement in order to calculate the local statistics
+    Only the metrics required by data privacy (such as count) or individual metrics of interested need to implement
+
+"""
+
+
+class Statistics(FLComponent, ABC):
+    def initialize(self, parts: dict, fl_ctx: FLContext):
+        """
+        Args:
+            parts: parts: components to be used by the Statistics
+            fl_ctx: fl_ctx: FLContext of the running environment
+
+        Returns:
+
+        """
+
+        pass
+
+    @abstractmethod
+    def features(self) -> Dict[str, List[Feature]]:
+        """
+           return Features for each dataset.
+
+           For example, we have training and test datasets.
+           the method will return
+           { "train": features1, "test": features2}
+           where features1,2 are the list of Features with contains feature name and DataType
+
+        Returns:Dict[<dataset_name>, List[Feature]]
+
+        Raises:
+            NotImplementedError
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def count(self, dataset_name: str, feature_name: str) -> int:
+        """
+           return record count for given dataset and feature
+           to perform data privacy min_count check, count is always required
+        Args:
+            dataset_name:
+            feature_name:
+
+        Returns: number of total records
+
+        Raises:
+            NotImplementedError
+
+        """
+        raise NotImplementedError
+
+    def sum(self, dataset_name: str, feature_name: str) -> float:
+        """
+            calculate local sums for given dataset and feature
+        Args:
+            dataset_name:
+            feature_name:
+        Returns: sum of all records
+
+        Raises:
+            NotImplementedError will be raised when sum metric is configured but not implemented. If the sum is not
+            configured to be calculated, no need to implement this method and NotImplementedError will not be raised.
+
+        """
+        raise NotImplementedError
```

Comment:
I think we should decorate this method with "@abstractmethod".
If they don't need this they can use "pass".


If want to do it another way, could be that make a metrics abstract class and these operation inherit from that class.
Instead of treating metrics method as FIXED function in this spec class.
That way users can add their own metrics easily later on.